ledger/ledger-runner-common/src/main/scala/com/daml/ledger/runner/common/PureConfigReaderWriter.scala

```diff
@@ -0,0 +1,343 @@
+
+import com.daml.ledger.api.tls.TlsVersion.TlsVersion
+import com.daml.ledger.api.tls.{SecretsUrl, TlsConfiguration, TlsVersion}
+import com.daml.lf.data.Ref
+import com.daml.lf.engine.EngineConfig
+import com.daml.lf.language.LanguageVersion
+import com.daml.lf.transaction.ContractKeyUniquenessMode
+import com.daml.lf.{VersionRange, interpretation, language}
+import com.daml.metrics.MetricsReporter
+import com.daml.platform.apiserver.SeedService.Seeding
+import com.daml.platform.apiserver.configuration.RateLimitingConfig
+import com.daml.platform.apiserver.{ApiServerConfig, AuthServiceConfig}
+import com.daml.platform.configuration.{
+  CommandConfiguration,
+  IndexServiceConfig,
+  InitialLedgerConfiguration,
+  PartyConfiguration,
+}
+import com.daml.platform.indexer.ha.HaConfig
+import com.daml.platform.indexer.{IndexerConfig, IndexerStartupMode}
+import com.daml.platform.services.time.TimeProviderType
+import com.daml.platform.store.DbSupport.{
+  ConnectionPoolConfig,
+  DataSourceProperties,
+  ParticipantDataSourceConfig,
+}
+import com.daml.platform.store.LfValueTranslationCache
+import com.daml.platform.store.backend.postgresql.PostgresDataSourceConfig
+import com.daml.platform.store.backend.postgresql.PostgresDataSourceConfig.SynchronousCommitValue
+import com.daml.platform.usermanagement.UserManagementConfig
+import com.daml.ports.Port
+import com.typesafe.config.ConfigValueFactory
+import io.netty.handler.ssl.ClientAuth
+import pureconfig.configurable.{genericMapReader, genericMapWriter}
+import pureconfig.error.CannotConvert
+import pureconfig.generic.semiauto._
+import pureconfig.{ConfigConvert, ConfigCursor, ConfigReader, ConfigWriter, ConvertHelpers}
+
+import scala.concurrent.duration.{Duration, FiniteDuration}
+import scala.jdk.DurationConverters.{JavaDurationOps, ScalaDurationOps}
+import scala.util.Try
+
+object PureConfigReaderWriter {
+  val Secret = "<REDACTED>"
+
+  /** Reads configuration object of `T` and `enabled` flag to find out if this object has values.
+    */
+  def optReaderEnabled[T](reader: ConfigReader[T]): ConfigReader[Option[T]] =
+    (cursor: ConfigCursor) =>
+      for {
+        objCur <- cursor.asObjectCursor
+        enabledCur <- objCur.atKey("enabled")
+        enabled <- enabledCur.asBoolean
+        value <-
+          if (enabled) {
+            reader.from(cursor).map(x => Some(x))
+          } else {
+            Right(None)
+          }
+      } yield value
+
+  /** Writes object of `T` and adds `enabled` flag for configuration which contains value.
+    */
+  def optWriterEnabled[T](writer: ConfigWriter[T]): ConfigWriter[Option[T]] = {
+    import scala.jdk.CollectionConverters._
+    def toConfigValue(enabled: Boolean) =
+      ConfigValueFactory.fromMap(Map("enabled" -> enabled).asJava)
+    (optValue: Option[T]) =>
+      optValue match {
+        case Some(value) =>
+          val configValue = writer.to(value)
+          if (configValue.atPath("temp").hasPath("temp.enabled")) {
```

Comment:
This looks strange (`temp.enabled`). Can you explain in a comment what's the idea here?

Reply:
I have reimplemented it differently, should be more clear I suspect :)